Online store

stankasvecova opened the pull request:
Version_2 of the online store.

stankasvecova committed:
README

stankasvecova committed:
version_1 (without display:grid)

stankasvecova committed:
version_2 (with display:grid, but not working)

BTBTravis commented on online_store/gt_sport_v1.html:
```diff
@@ -0,0 +1,120 @@
+<script>
+const sportProducts = [
+  {
+    imgSrc: "skateboard.jpg",
```
where are these images? Did you add commit them?
<img width="351" alt="Screenshot 2020-04-30 at 13 08 58" src="https://user-images.githubusercontent.com/13404719/80703847-d3922100-8ae3-11ea-9c31-f0b213c38afe.png">

stankasvecova replied:
Re-uploaded them. Probably deleted them by mistake when 'tidying up'.

BTBTravis commented on online_store/gt_sport_v1.html:
```diff
@@ -0,0 +1,120 @@
+    <title>goodTeens sports online store</title>
+  </head>
+  <body>
+<h2 class="column" id="h2title">goodTeens - Sports</h2> <p>
```
why are you opening a paragraph tag at the end of this line?

stankasvecova replied:
True, a mistake, got rid of it.

BTBTravis commented on online_store/gt_sport_v1.html:
```diff
@@ -0,0 +1,120 @@
+    <link rel="stylesheet" href="gt_sports.css">
+    <title>goodTeens sports online store</title>
+  </head>
+  <body>
```
There are indentation problems in you HTML and JavaScript. I would suggest two things to clean it up. 

1. use a separate js file for all your javascript
1. run vs codes auto formatted on both this html and the newly created js file. 

I created a video explaining how to use VS Code auto format [here](https://www.youtube.com/watch?v=QeGp3wKU3Jg) with this exact code as an example.

stankasvecova replied:
The video is amazing and fixed my indentation problems, thanks!
I have now also a separate .js file. Great suggestiongs, thanks.

BTBTravis commented on online_store/gt_sport_v1.html:
```diff
@@ -0,0 +1,120 @@
+    seller: "Oceana",
+  },
+  {
+    imgSrc: "volleyball.jpg",
```
I believe part of the homework was to use a nested object. Perhaps we could nest the image src and include alt text.

```js
{
    img: {
        source: "volleyball.jpg",
        altText: "volleyball image"
    },
   name: "Stand-up paddle",
...
```

stankasvecova replied:
In my version_2, I have nested objects already, so unless this is necessary and requested, I wouldn't add another. Are you OK with that?

BTBTravis commented on online_store/gt_sport_v2.html:
```diff
@@ -0,0 +1,119 @@
+];
+
+function productDescriptionDisplay() {
+  for (let i=0; i < sportProducts.length; i++) {
```
this for loop is never closed. you need to add an `}` somewhere to close it. Correct indention makes this bug easier to spot

stankasvecova replied:
Yep, found it, auto format document helped! Fixed! Thx

BTBTravis commented on online_store/gt_sport_v2.html:
```diff
@@ -0,0 +1,119 @@
+};
+
+function productImgDisplay() {
+  for (let i=0; i < sportProducts.length; i++) {
```
another unclosed for loop.

BTBTravis reviewed (changes requested):
Hey @stankasvecova  great work so far. I've added some comment of things to update.

stankasvecova committed:
Add files via upload

stankasvecova committed:
version2

Style grid is not working
Javascript is not creating new <ul> or <div> (tried both) and is not loading content for all the products

stankasvecova committed:
clean-up

BTBTravis commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+
+online_store/gt_sport_v1.html
```
If you want to remove the gt_sport_v1 files i would suggest just deleting them and committing that change. You can always bring them back later from the git history.

stankasvecova replied:
Deleted.

BTBTravis commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+
```
So git ignore only prevents future files from being tracked by git. It does not do anything to files already tracked by the repo. To remove a file from being tracked you can either delete it and commit that delete or `git rm --cached /path/to/file` which deletes from file from git but not your working directory.

stankasvecova replied:
Cool, thx.

BTBTravis commented on online_store/gt_sport_v2.js:
```diff
@@ -0,0 +1,92 @@
+      <p> This ${sportProducts[i].name} is suitable for teenagers on ${sportProducts[i].level} level.
+ `;
+    productDescription.appendChild(descriptions);
+    descriptions.innerHTML = productDescriptionDisplay();
```
⚠️ this is a recursive call and is not what you want it will run indefinitely

stankasvecova replied:
Yep, deleted! Great job spotting!

stankasvecova committed:
Delete gt_sport_v1.html

stankasvecova committed:
Delete gt_sport_v1.css

stankasvecova committed:
updated

stankasvecova committed:
Delete lecturework.html

stankasvecova committed:
Merge branch 'online_store' of https://github.com/stankasvecova/redi_js_spring_2020 into online_store

stankasvecova committed:
updates

stankasvecova committed:
Shopping Cart & Total Price

MrTim commented on online_store/gt_sport_v2.js:
```diff
@@ -0,0 +1,122 @@
+};
+productDescriptionDisplay();
+
+const btnEl = productEl.querySelector("button");
```
productEl is an array, that's why you can't use `querySelector` on it. Probably you meant document.

stankasvecova replied:
Yep.

MrTim commented on online_store/gt_sport_v2.js:
```diff
@@ -0,0 +1,122 @@
+productDescriptionDisplay();
+
+const btnEl = productEl.querySelector("button");
+btnEl.addEventListener('click', addProductsToCart);
```
this line would fail because you don't have `addProductsToCart`  function

stankasvecova replied:
Aha. Thanks for noticing. I had following code there before:
`function addProductsToCart() {
 for (let i = 0; i < sportProducts.length; i++) {
 shoppingCart.push(sportProducts[i]);
 };
 console.log(shoppingCart);
};`
but got rid of it by mistake...

MrTim commented on online_store/gt_sport_v2.js:
```diff
@@ -0,0 +1,122 @@
+
+const btnEl = productEl.querySelector("button");
+btnEl.addEventListener('click', addProductsToCart);
+productListEl.appendChild(productEl);
```
`productListEl` is not defined as well.

stankasvecova replied:
Fixed and defined.

MrTim commented on online_store/gt_sport_v2.js:
```diff
@@ -0,0 +1,122 @@
+      <p> Sold by a <b>gT partner</b> called ${sportProducts[i].seller}.</p>
+      <p> This ${sportProducts[i].name} is suitable for teenagers on ${sportProducts[i].level} level.
+      <p> <button type="button" id="addToShoppingCartBtn">Add to Shopping Cart</button>
+      `;
```
I recommend you to do the following if you want to add a button to each product.
```javascript
const btnEl = document.createElemen("button");
btnEl.addEventListener('click', function() {
  addProductToCart(sportProducts[i]);
});
```
This would give you possibility when button was pressed pass object (sport product) to the function which is actually adding them to your shopping cart.

stankasvecova replied:
Added, thanks.

BTBTravis replied:
This was good advice 👌

MrTim commented on online_store/gt_sport_v2.js:
```diff
@@ -0,0 +1,122 @@
+function displayContentInCart() {
+  for (let i = 0; i < shoppingCart.length; i++) {
+    const contentEl = document.getElementById("shoppingCart")
+    productEl.innerHTML = "";
```
you don't need this line

stankasvecova replied:
...happy to get rid of it, but wasn't it meant to clear my shoppingCart before it adds things into it?

MrTim commented on online_store/gt_sport_v2.js:
```diff
@@ -0,0 +1,122 @@
+
+function calculateTotalPrice() {
+  let totalPrice=0;
+  for (let i=0; i < shoppingCart.length; i++){
```
space around `=`

stankasvecova replied:
fixed

MrTim commented on online_store/gt_sport_v2.js:
```diff
@@ -0,0 +1,122 @@
+  alert(totalPrice);
+};
+
+shoppingCartEl.addEventListener(displayContentInCart);
```
addEventListener has 2 arguments: first one is event name and second one is your function (callback)

stankasvecova replied:
a typo. Of course!
`shoppingCartEl.addEventListener("click", displayContentInCart);`

MrTim reviewed (changes requested):
Good work! You're pretty close to the final result. I left some comments which should help you to move forward.

stankasvecova committed:
Fixed some bugs

caritosteph commented on online_store/gt_sport_v2.js:
```diff
@@ -0,0 +1,129 @@
+    contentEl.appendChild(productEl);
+  };
+};
+displayContentInCart();
```
the function shouldn't be called here, just when user clicks ShowShoppingCart button.

stankasvecova replied:
Thnx, deleted

MrTim commented on online_store/gt_sport_v2.js:
```diff
@@ -110,13 +117,13 @@ function displayContentInCart() {
 };
 
-shoppingCartEl.addEventListener(displayContentInCart);
+shoppingCartEl.addEventListener("click", displayContentInCart);
```
It's not shoppingCartEl - because it's not a button, I think you want to use button with id='intoSC'.

stankasvecova replied:
changed.

MrTim commented on online_store/gt_sport_v2.js:
```diff
@@ -91,15 +91,22 @@ function productDescriptionDisplay() {
+  console.log(shoppingCart);
+};
+
+const btnEl = document.createElement("button");
```
this button should be created not in here, but right after the product description, inside you're for loop.
addProductsToCart - should be a function with one parameter. and it should looks somethign like that:
```javascript
function addProductsToCart(product) {
    shoppingCart.push(product);
};
```
Otherwise this function (as it is right now) would add all element to shopping cart each time you click on any button, because you have loop inside.

stankasvecova replied:
Changed, done. Thx.

MrTim reviewed (changes requested):
Commented on some points. You can adjust those and continue.

stankasvecova committed:
debugged further

caritosteph commented on online_store/gt_sport_v2.js:
```diff
@@ -71,11 +71,16 @@ let productEl = [];
+let shoppingCartEl = document.getElementById("shoppingCart");
+let intoSCbtn = document.getElementById("intoSC");
+let calTPbtn = document.getElementById("calTP");
+let addToShoppingCartBtn = document.getElementById("addToShoppingCartBtn");
```
Here it is getting the button element before it is being rendered.
Below you have the function **productDescriptionDisplay** which display the list of product with each button, so after that you have access to your button element.

stankasvecova replied:
Sounds good

caritosteph commented on online_store/gt_sport_v2.js:
```diff
@@ -87,21 +92,22 @@ function productDescriptionDisplay() {
-  for (let i = 0; i < sportProducts.length; i++) {
-    shoppingCart.push(sportProducts[i]);
-  };
+  shoppingCart.push(sportProducts[i]);
```
you can not use **i** before it is being declared.

stankasvecova replied:
OK, changed.

caritosteph commented on online_store/gt_sport_v2.js:
```diff
@@ -87,21 +92,22 @@ function productDescriptionDisplay() {
       <p> <button type="button" id="addToShoppingCartBtn">Add to Shopping Cart</button>
```
**Recommendation**
Here you should generate different id for each button because id are always unique.
for the problem that you are facing have 2 solution.
_Solution 1:_
generate different id for each button like: addToShoppingCartBtn0, addToShoppingCartBtn1, etc.
then after you render your component you can addEventListener **click** to each button.
_Solution 2:_
you can read a bit about [onclick](https://www.w3schools.com/tags/ev_onclick.asp) event and use it in your button, this way you don't need to add ids.

stankasvecova replied:
Thanks, I will go with 'onclick' event.

stankasvecova committed:
updates

BTBTravis commented on online_store/gt_sport_v2.js:
```diff
@@ -0,0 +1,139 @@
+
+      const addToShoppingCartBtn = document.createElement("button");
+      addToShoppingCartBtn.textContent = "Add to shopping Cart";
+      addToShoppingCartBtn.addEventListener("click", addProductsToCart);
```
Here we have a problem because we are passing the `addProductsToCart` function as a callback to the `addEventListener`. The `addEventListener` is expecting the callback to be a function that gets the DOMEvent. So we end up adding dom events to the cart not products 🙃. 

We could solve this in many ways but how about creating an anonymous function as the callback function and inside that new anonymous fn we call addProductsToCart with the product we already have in scope. I'm not adding the code because I want to challenge you. 💪

stankasvecova replied:
Ok, would this work?
`    addToShoppingCartBtn.addEventListener('click', function () {
      shoppingCart.push(product);
};`

stankasvecova replied:
solved :)

BTBTravis reviewed (changes requested):
Let's fix the add to cart functionality

stankasvecova committed:
work in progress

stankasvecova committed:
work in progress

stankasvecova committed:
updates

stankasvecova committed:
added availability of products

stankasvecova committed:
fixed bugs

stankasvecova commented:
My last trouble is:
the shoppingCart is not being displayed properly, it says: 
undefined: undefined. 
Help, anyone?
Thanks

BTBTravis commented on online_store/gt_sport_v2.js:
```diff
@@ -0,0 +1,182 @@
+    let cartList = document.createElement("div");
+    intoSCbtn.addEventListener("click", function () {
+      cartList.innerHTML = `
+      ${shoppingCart.name}:
```
shoppingCart is an array and does not have a name property. Perhaps you meant to access an item of the array?

stankasvecova replied:
I discussed it with Ivan, I had the same question, but because shoppingCart takes identically structured data from sportProducts, then apparently it should have .name and any other properties that sportProducts have...what do you think?

stankasvecova replied:
resolved based on your suggestion, thanks.

BTBTravis commented on online_store/gt_sport_v2.js:
```diff
@@ -0,0 +1,182 @@
+      <p> Last chance! Only 1 more left in stock! </p>
+      `;
+    }
+    else if (sportProducts[i].availability > 1 && sportProducts[i].availability <= 3) {
```
each of these if blocks has a ton of duplication. perhaps you could use a function or different structuring to repeat yourself less.

stankasvecova replied:
Thanks

caritosteph commented on online_store/gt_sport_v2.js:
```diff
@@ -0,0 +1,182 @@
+function displayContentInCart() {
+  for (let i = 0; i < shoppingCart.length; i++) {
+    let cartList = document.createElement("div");
+    intoSCbtn.addEventListener("click", function () {
```
this addEventLister is not needed here.
the function displayContentInCart is being call when user clicks in **intoSCbtn**.

stankasvecova replied:
Thanks, implemented.

stankasvecova committed:
updates...

stankasvecova committed:
final version

BTBTravis reviewed (approved):
Looks great thanks for you hard work

stankasvecova committed:
the final tweaks

MrTim commented on online_store/gt_sport_v2.js:
```diff
@@ -157,11 +138,12 @@ function addProductsToCart(product) {
 
 function displayContentInCart() {
   for (let i = 0; i < shoppingCart.length; i++) {
```
Before for loop you need to clean the innerHTML of shoppingCartEl. Otherwise you duplicate content inside.

stankasvecova replied:
Yep, thanks, didn't notice!
Fixed now!

MrTim commented on online_store/gt_sport_v2.js:
```diff
@@ -177,3 +159,16 @@ function calculateTotalPrice() {
+  const emptyStars = maxStars - rating;
+  //return "Rating:  ★ ☆ ☆ ☆ ☆";
+  let starText = "";
+  for (let i=0; i < rating; i++) {
```
spaces around `=`

stankasvecova replied:
Done

MrTim reviewed (changes requested):
Really cool! I like it!
Please fix one small bug regarding shopping cart and after that PR would be approved)

stankasvecova committed:
Final version

MrTim reviewed (approved):
Great job! Could be merged to master.

stankasvecova committed:
final version
